worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala

```diff
@@ -245,6 +247,11 @@ class PushDataHandler(val workerSource: WorkerSource) extends BaseMessageHandler
       fileWriter.decrementPendingWrites()
       return;
     }
+    // Write to local async.
+    // If write to buffer fails, respond back to client with failure. Else proceed with replication.
+    if (!writeDataWithExceptionHandling(fileWriter, body, shuffleKey, callbackWithTimer)) {
```

Comment:
I don't think we should do this before data is pushed to replica asynchronously. This adds a delay to start of replication which could be significant when append does involve a flush.

Reply:
Agree with @otterc, instead we should parallelize both operations - and fail if either of them fail, and return success only if both do.
Something vaguely like:
```
val localWriteFuture = Future {
  writeLocally()
}

val remoteFuture = Future {
  replicateToPeer()
}

Future.sequence(Seq(localWriteFuture, remoteFuture)).onComplete {
  case Success(_) => sendSuccessToCaller()
  case Failure(_) => reportFailure()
}

```

Reply:
fixed

Reply:
> I don't think we should do this before the data push to peer. This adds a delay to start of replication which could be significant when append does involve a flush.

Hi @otterc , just for one small clarification, flush is always done in `Flusher`'s dedicated threads, `handlePushData/handlePushMergedData` will not do real flush :)